### ISSUE TITLE
Incompatible with Micronaut 4.2

### ISSUE THREAD

**Issue by alexmartin-b1** [database-rider/database-rider#577]
I'm working to upgrade my projects from Micronaut 3 to Micronaut 4 and have encountered an issue where it cannot retrieve a database connection. I've spun up a small example project which pulls in the relevant micronaut data projects as well as DB Rider and a base test that should seed the database but it is failing. 

Demo Repository: https://github.com/alexmartin-b1/dbrider-micronaut-demo

Stacktrace:
```
No current connection present. Consider declaring @Connectable or @Transactional on the surrounding method
io.micronaut.data.connection.exceptions.NoConnectionException: No current connection present. Consider declaring @Connectable or @Transactional on the surrounding method
	at app//io.micronaut.data.connection.exceptions.NoConnectionException.notFoundInAdvice(NoConnectionException.java:49)
	at app//io.micronaut.data.connection.jdbc.advice.ContextualConnectionInterceptor.intercept(ContextualConnectionInterceptor.java:58)
	at app//io.micronaut.aop.chain.MethodInterceptorChain.proceed(MethodInterceptorChain.java:137)
	at app//io.micronaut.data.connection.jdbc.advice.ContextualConnection$Intercepted.getMetaData(Unknown Source)
	at app//com.github.database.rider.core.util.DriverUtils.getDriverName(DriverUtils.java:48)
	at app//com.github.database.rider.core.connection.RiderDataSource.resolveDBType(RiderDataSource.java:184)
	at app//com.github.database.rider.core.connection.RiderDataSource.checkDbType(RiderDataSource.java:176)
	at app//com.github.database.rider.core.connection.RiderDataSource.init(RiderDataSource.java:97)
	at app//com.github.database.rider.core.connection.RiderDataSource.<init>(RiderDataSource.java:54)
	at app//com.github.database.rider.core.dataset.DataSetExecutorImpl.getRiderDataSource(DataSetExecutorImpl.java:924)
	at app//com.github.database.rider.core.RiderRunner.setup(RiderRunner.java:30)
	at app//com.github.database.rider.junit5.DBUnitExtension.beforeTestExecution(DBUnitExtension.java:68)
	at java.base@17.0.3/java.util.ArrayList.forEach(ArrayList.java:1511)
	at java.base@17.0.3/java.util.ArrayList.forEach(ArrayList.java:1511)
	```

I included a separate test in `DemoEntityTest` which is outside of DB Rider which shows that we are able to both persist and assert on data stored in postgres.

**Comment by dominictobias** [comment]
I see that the issues in the milestones are all done, are there still outstanding issues to make this work? Would be great to have compat with v4

**Comment by rmpestano** [comment]
Hey @dominictobias, I couldn't look into that yet, I'll try to look into it soon, thanks for the demo

**Comment by gnarroway-bullish** [comment]
FWIW, I see basically the same issue going from micronaut 3.5 to [3.6.x](https://micronaut.io/2022/08/04/micronaut-framework-3-6-0-released/), so I suspect it was introduced a lot earlier than 4.x

The closest relevant discussions I've seen are:

- https://stackoverflow.com/questions/66472191/micronaut-db-scheduler-no-current-transaction-present-consider-declaring-tra/71933329#71933329
- https://github.com/micronaut-projects/micronaut-data/issues/1083

That said, those discussions pre-date micronaut 3.6 by a year

Tried with dbrider 1.44.0

**Comment by gnarroway-bullish** [comment]
I have fixed @alexmartin-b1 's demo project [in a fork](https://github.com/gnarroway-bullish/dbrider-micronaut-demo), with several commits to show the evolution:

1. Establish a working baseline with micronaut 3.5.3 (https://github.com/gnarroway-bullish/dbrider-micronaut-demo/commit/5bc0d33cfafbcc79f72b7a310ace4f81d8243ed3)
    - this is to establish a known good state that works out of the box
    - note it does not work if `@MicronautTest(transactional=false)` (the default is true)
2. Simply update the micronaut version to 3.6 to demonstate it breaks (https://github.com/gnarroway-bullish/dbrider-micronaut-demo/commit/aec60391f57df1f672cc7a5f28d1571d1dc61e41)
    - presumably something changed in the way it wraps the connection
3. Fix it for micronaut 3.6 (https://github.com/gnarroway-bullish/dbrider-micronaut-demo/commit/5caac9de49c0643fc18cbe2e4c23a449abeb9148)
    - this involves unwrapping the connection before using it in dbrider, specifically by providing our own 
`ConnectionManager` implementation with just 2 lines changed:
    
    ```
    DataSource unwrappedDataSource = DelegatingDataSource.unwrapDataSource(dataSource);
    return new ConnectionHolderImpl(unwrappedDataSource.getConnection());
    ```
    - note this happens to now fix it even when `@MicronautTest(transactional=false)`
4. Finally, make it work with the original micronaut 4.2 example (https://github.com/gnarroway-bullish/dbrider-micronaut-demo/commit/056d8a59f7c21cad6590facf47c8a811c8e60bb1)
    - the strategy is the same as the preceding commit but the package has moved

So in summary, we need to unwrap the datasource for it to function correctly in micronaut, otherwise it is looking for a transaction that is not there.

**Comment by rmpestano** [comment]
Amazing work @gnarroway-bullish, can you open the PR?

**Comment by vlusslus** [comment]
Hello, can you please add support for Micronaut 4 via PR?
